atomic_reactor/plugins/pre_resolve_module_compose.py

```diff
@@ -0,0 +1,272 @@
+            else:
+                self.log.debug("Retrying request compose_id={}, elapsed_time={}"
+                               .format(compose_id, elapsed))
+
+                if elapsed > burst_length:
+                    time.sleep(slow_retry)
+                else:
+                    time.sleep(burst_retry)
+
+
+class ResolveModuleComposePlugin(PreBuildPlugin):
+    key = "resolve_module_compose"
+    is_allowed_to_fail = False
+
+    def __init__(self, tasker, workflow,
+                 module_name, module_stream, module_version=None,
+                 compose_id=None,
+                 odcs_url=None, odcs_insecure=False,
+                 odcs_openidc_secret_path=None,
+                 pdc_url=None, pdc_insecure=False):
+        """
+        constructor
+
+        :param tasker: DockerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param module_name: Module name to look up in PDC
+        :param module_stream: Module stream to look up in PDC
+        :param module_version: Module version to look up in PDC (optional)
+        :param compose_id: ID of compose in ODCS (optional - will only be set for workers)
+        :param odcs_url: URL of ODCS (On Demand Compose Service)
+        :param odcs_insecure: If True, don't check SSL certificates for `odcs_url`
+        :param odcs_openidc_secret_path: directory to look in for a `token` file (optional)
+        :param pdc_url: URL of PDC (Product Definition center))
+        :param pdc_insecure: If True, don't check SSL certificates for `pdc_url`
+        :
+        """
+        # call parent constructor
+        super(ResolveModuleComposePlugin, self).__init__(tasker, workflow)
+
+        if not pdc_url:
+            raise RuntimeError("pdc_url is required")
+        if not odcs_url:
+            raise RuntimeError("odcs_url is required")
+        self.module_name = module_name
+        self.module_stream = module_stream
+
+        if module_version is not None and re.match(r'^\d{14}$', module_version) is None:
+            raise RuntimeError("module_version should be 14 digits")
+        self.module_version = module_version
+
+        self.compose_id = compose_id
+        self.odcs_url = odcs_url
+        self.odcs_insecure = odcs_insecure
+        self.odcs_openidc_secret_path = odcs_openidc_secret_path
+        self.pdc_url = pdc_url
+        self.pdc_insecure = pdc_insecure
+
+    def _resolve_compose(self):
+        if self.odcs_openidc_secret_path:
+            token_path = os.path.join(self.odcs_openidc_secret_path, 'token')
+            with open(token_path, "r") as f:
+                odcs_token = f.read().strip()
+        else:
+            odcs_token = None
+
+        odcs_client = ODCSClient(self.odcs_url, insecure=self.odcs_insecure, token=odcs_token,
+                                 log=self.log)
+        pdc_client = PDCClient(server=self.pdc_url, ssl_verify=not self.pdc_insecure, develop=True)
```

Comment:
`develop` is for using in a development environment? Should this be a module parameter?

Reply:
'develop' actually means "try API requests without authentication" - since all the PDC access we need is public and possible without authentication, this makes things a lot simpler and avoids us having to set up OSBS <=> PDC authentication. Clearly needs a comment.

Reply:
Looks like this still needs a comment.

Reply:
Sorry, missed that - added, will be in my next push.